PHP 8.4 support: Update Doctrine and PHPStan

costasovo opened the pull request:
## Description

This PR updates Doctrine and PHPStan. 
This addresses PHP warnings about implicitly nullable params coming from Doctrine proxy classes for entities, and it is also possible to run PHPStan (`./do qa:phpstan`) on PHP 8.4.

## Code review notes

- For testing commands `./do qa:phpstan` and `./do doctrine:generate-cache` on PHP 8.4, run `./do install` first.
- There is a change/regression in the newer Doctrine that `$repo->findBy(['id' => $listOfIds])` doesn't work for empty arrays (https://github.com/doctrine/orm/issues/12245). We didn't have many occurrences, so I replaced them with an alternative approach.
- I updated PHPStan to the latest 1.x version. We should be able to update to v2 as it supports PHP7.4, but for time reasons, I updated only to the latest v1.

## QA notes

Please run some basic smoke tests. There's a change in code that generates SQL when sending, so please test sending to a larger batch (100+ recipients).

## Linked PRs

_N/A_

## Linked tickets

STOMAIL-7513

## Tasks

- [ ] I have added a changelog entry (`./do changelog:add --type=<type> --description=<description>`)
- [ ] I followed [best practices](https://codex.wordpress.org/I18n_for_WordPress_Developers) for translations
- [ ] I added sufficient test coverage
- [ ] I embraced TypeScript by either creating new files in TypeScript or converting existing JavaScript files when making changes

## Preview

[Preview in WordPress Playground](https://account.mailpoet.com/playground/new/branch:stomail-7513-fix-phpstan-and-doctrine-cache-in-php-84)

_The latest successful build from `stomail-7513-fix-phpstan-and-doctrine-cache-in-php-84` will be used. If none is available, the link won't work._

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added a utility to auto-fix PHP 8.4 deprecations and accompanying unit tests.

* **Bug Fixes**
  * More robust bulk-ID lookups across segments/forms/subscribers to improve data retrieval.
  * Safer newsletter/send queue and subscription handling (status/typing/encoding guards).
  * Hardened analytics aggregation and import/export flows to avoid runtime notices.

* **Chores**
  * Updated static analysis baselines, CI bootstrap scripts and related dev-tooling/dependency updates.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

costasovo committed:
Update Doctrine dbal, common and orm + subdependencies

STOMAIL-7513

costasovo committed:
Update doctrine cleanup/fix script

STOMAIL-7513

costasovo committed:
Add fix for PHP 8.4 Deprecations also to test_env

costasovo committed:
Update PHPStan to latest v1 packages

costasovo committed:
Update doctrine/orm to latest v2 patch

costasovo committed:
Point PHPStan to dependencies installed in vendor-prefixed

PHPStan for Doctrine needs to know the version of the Doctrine packages.
We need to give the composer the correct data to return the proper versions.

STOMAIL-7513

costasovo committed:
Replace deprecated settings with their error indetifier alternatives

New version of PHPStan started reporting deprecation warnings and
suggested this edit.

costasovo committed:
Fix PHPStan warnigns that surfaced in lib after update

costasovo committed:
Address new PHPStan errors in tests

costasovo committed:
Regenerate baseline files

After all warnings were fixed I regenerated the baseline files
to get rid of "unmatched error pattern" warnings.

STOMAIL-7513

costasovo committed:
Remove unnecessary constant definition

costasovo committed:
Fixes for doctrine removed support for findBy(['id' => array()]);

costasovo committed:
Adjust PHPStan number of processes to prevent memory issues in Circle CI

costasovo committed:
Update PHP4 Deprecation fix based on core rabbit feedback

The current pattern only matches parameters shaped like Type $param = null. It skips legitimate signatures such as function foo(Bar &$ref = null) or function foo(Bar ...$args = null), so those parameters remain implicitly nullable and still trigger the PHP 8.4 deprecation you’re trying to suppress. Typed reference parameters with default null are valid PHP, so the fixer needs to cover them too.

This also refactors code duplication.

costasovo commented on mailpoet/lib/Mailer/WordPress/WordPressMailer.php:
```diff
@@ -79,7 +79,13 @@ private function getEmail() {
+        $bodyForConversion = $converted !== false ? $converted : $this->Body;
+      }
+      $text = @Html2Text::convert($bodyForConversion);
       $email['body']['text'] = $text;
```
I addressed it by changing to the previous code where we had no fallback condition branch. Actually, we don't work further with information whether the conversion was successful or not, so the output provided when the 'pass' charset is matched works for us in the same way as the fallback.

costasovo commented on mailpoet/lib/Subscription/Manage.php:
```diff
@@ -183,7 +183,7 @@ private function updateSubscriptions(SubscriberEntity $subscriber, array $subscr
     }
 
     if ($subscriber->getStatus() === SubscriberEntity::STATUS_SUBSCRIBED && $newSegmentIds) {
-      $newSegments = $this->segmentsRepository->findBy(['id' => $newSegmentIds]);
+      $newSegments = $this->segmentsRepository->findByIds($newSegmentIds);
       $this->newSubscriberNotificationMailer->send($subscriber, $newSegments);
```
I added array_map('intval', $newSegmentIds) directly to the findByIds method.

costasovo commented on mailpoet/tasks/Php84DeprecationsFix.php:
```diff
@@ -0,0 +1,105 @@
+    }
+  }
+
+  private function fixImplicitlyNullableParameters(string $content): string {
+    // Find potential matches
+    $pattern = '/(\s|,|\()\s*([A-Za-z_\\\\][A-Za-z0-9_\\\\]*)(\s+&?\s*(?:\.\.\.)?\$\w+\s*=\s*(?:null|NULL)\b)/';
+
+    $updated = preg_replace_callback($pattern, function($matches) {
+      $fullMatch = $matches[0];
+      $prefix = $matches[1];
+      $type = $matches[2];
+      $paramPart = $matches[3];
+
+      // Additional safety checks
+      if (strpos($fullMatch, '|') !== false) {
+        return $fullMatch; // Don't modify union types
+      }
+
+      // Check if already nullable (look for ? before the type in the prefix)
+      if (strpos($prefix, '?') !== false) {
+        return $fullMatch; // Already nullable
+      }
+
+      // Don't modify visibility modifiers or storage class specifiers
+      $excludedKeywords = ['static', 'private', 'protected', 'public', 'const', 'var'];
+      if (in_array(trim($type), $excludedKeywords)) {
+        return $fullMatch;
+      }
+
+      // Don't modify 'mixed' type since it already includes null
+      if (trim($type) === 'mixed') {
+        return $fullMatch;
+      }
+
+      return $prefix . '?' . $type . $paramPart;
+    }, $content);
+
+    return $updated ?? $content;
+  }
```
The PHP 8.4 fix is temporary and we don't run it on production code. We will update dependencies and remove this fix in the future. So for now this is acceptable.

NeosinneR commented on mailpoet/tasks/Php84DeprecationsFix.php:
```diff
@@ -0,0 +1,105 @@
+      // Check if already nullable (look for ? before the type in the prefix)
+      if (strpos($prefix, '?') !== false) {
+        return $fullMatch; // Already nullable
+      }
```
I did ask multiple agents to analyze `Php84DeprecationsFix.php` file, and all pointed out to this condition as being broken. I haven't checked if this was present before or was introduced with changed `$pattern`.

**Grok**
<img width="551" height="108" alt="Screenshot 2025-11-03 at 15 49 50" src="https://github.com/user-attachments/assets/177f78bc-f3c7-4b41-83f2-133746f3e969" />

**Claude**
<img width="589" height="110" alt="Screenshot 2025-11-03 at 15 50 26" src="https://github.com/user-attachments/assets/01a73858-ae3b-4e72-915e-6232306fedd1" />

**Cursor**
<img width="660" height="104" alt="Screenshot 2025-11-03 at 15 51 02" src="https://github.com/user-attachments/assets/7c45ae0c-b891-4216-bff7-21f00beb1ea2" />

costasovo replied:
Hmm, it could be related to the pattern update. It was suggested by CodeRabbit. I'll look into it.

NeosinneR reviewed (commented):
I switched to PHP 8.4 and see this error when running `./do qa:phpstan` or `./do doctrine:generate-cache`. Have you seen this, or did I just break my local env? 😄 

<img width="1530" height="799" alt="Image" src="https://github.com/user-attachments/assets/50f33f86-afb0-434d-985c-5e9e6f04267e" />

costasovo commented:
> I switched to PHP 8.4 and see this error when running ./do qa:phpstan or ./do doctrine:generate-cache. Have you seen this, or did I just break my local env? 

@NeosinneR Did you run `./do install` after switching to PHP8.4? The Php84DeprecationsFix.php has a condition to run only when the PHP version is 8.4. It always had... But thinking about it as I'm typing, do we need it? Can we just run it for whatever version? With the current implementation, the PHP84 fixes are applied only in the dev env, aren't they?

NeosinneR commented:
I saw that error also when `./do install`. The issue was that I only switched main Docker to PHP 8.4 and not the one in `tests_env`. Once I switched both, `./do install`, `./do qa:phpstan` and `./do doctrine:generate-cache` all succeeded. 

> do we need it? Can we just run it for whatever version?

I don't have strong preference. I don't really see issue in gating the script only on PHP 8.4, as it has no practical value running elsewhere.

costasovo commented:
> I saw that error also when ./do install. The issue was that I only switched main Docker to PHP 8.4 and not the one in tests_env. Once I switched both, ./do install, ./do qa:phpstan and ./do doctrine:generate-cache all succeeded.

@NeosinneR Makes sense. I was switching PHP on the host and ran ./do install from the host. So I didn't experience the issue with the test container.

> I don't really see issue in gating the script only on PHP 8.4, as it has no practical value running elsewhere.

I'm a bit confused because I thought it had practical value. Double check with me, please 🙏 . I see that it patches dependencies in vendor and vendor-prefixed folders. The vendor-prefixed folder contains only production dependencies, and the vendor also has some production dependencies. In CircleCI, we [use PHP 8.3 when we build the zip](https://github.com/mailpoet/mailpoet/blob/d32e32d7786cc864ad89a96e5128be73cebabbfc/.circleci/config.yml#L115) so that means these dependencies are not patched in the zip we actually release. So currently it looks like to me that we don't fix PHP84 issues for our users.

To my understanding, if we allow the Php84DeprecationsFix to run for all PHP versions, we would workaround the issue with having to run `./do install`, and we would start patching production dependencies.

NeosinneR commented:
@costasovo okay, that does make sense, and I do see now why we actually should remove that PHP 8.4 check and run the script on all version.

costasovo committed:
Fix PHP 8.4 deprecation fix condition and add a unit test

We don't have a special environment for testing tasks, so I placed
The test to unit tests.

costasovo committed:
Allow fixes for PHP 8.4 deprecation to run for all PHP versions

This is to prevent the need to reinstall everything after switching PHP versions
and to run those fixes also for the production build.
STOMAIL-7513

costasovo commented:
@NeosinneR I updated the PR. 
- allowed fixes for all PHP versions
- updated the incorrect condition in the Php84DeprecationsFix file and asked AI to write a unit test for it

NeosinneR reviewed (approved):
Thanks for the update. And 👍 for the tests.

Aschepikov reviewed (approved):
GJ!
